src/performance/metrics/StatisticCollector.cpp

```diff
@@ -46,64 +48,172 @@ int StatisticCollector::getMemoryUsageByProcess() {
         }
     }
     fclose(file);
-    std::cout << "Memory Usage: " + std::to_string(result) << std::endl;
+    // std::cout << "Memory Usage: " + std::to_string(result) << std::endl;
     return result;
 }
 
-int StatisticCollector::parseLine(char* line) {
-    int i = strlen(line);
-    const char* p = line;
-    while (*p < '0' || *p > '9') p++;
-    line[i - 3] = '\0';
-    i = atoi(p);
-    return i;
+int StatisticCollector::getThreadCount() {
+    FILE *file = fopen("/proc/self/stat", "r");
+    long result;
+    char line[128];
+
+    for (int i = 0; i < 20; i++) {
+        if (fscanf(file, "%127s%*c", line) < 0) {
+            fclose(file);
+            return -1;
+        }
+    }
+    fclose(file);
+    result = strtol(line, NULL, 10);
+    if (result <= 0 || result > 0xfffffffffffffffL) return -1;
+    return result;
 }
 
-double StatisticCollector::getCpuUsage() {
-    struct tms timeSample;
-    clock_t now;
-    double percent;
-
-    now = times(&timeSample);
-    if (now <= lastCPU || timeSample.tms_stime < lastSysCPU || timeSample.tms_utime < lastUserCPU) {
-        // Overflow detection. Just skip this value.
-        percent = -1.0;
-    } else {
-        percent = (timeSample.tms_stime - lastSysCPU) + (timeSample.tms_utime - lastUserCPU);
-        percent /= (now - lastCPU);
-        percent /= numProcessors;
-        percent *= 100;
+static long getSwapSpace(int field) {
+    FILE *file = fopen("/proc/swaps", "r");
+    long result = -1;
+    char line[128];
+
+    fgets(line, 128, file);
+
+    while (fgets(line, 128, file) != NULL) {
+        char *value;
+        char *save = NULL;
+        for (int i = 0; i < field; i++) {
+            if (i == 0) {
+                value = strtok_r(line, " ", &save);
+            } else {
+                value = strtok_r(NULL, "\t", &save);
+            }
+        }
+        long used = strtol(value, NULL, 10);
+        if (used < 0 || used > 0xfffffffffffffffL) {
+            continue;
+        }
+        if (result >= 0) {
+            result += used;
+        } else {
+            result = used;
+        }
     }
-    lastCPU = now;
-    lastSysCPU = timeSample.tms_stime;
-    lastUserCPU = timeSample.tms_utime;
+    fclose(file);
 
-    return percent;
+    return result;
 }
 
-std::string StatisticCollector::collectVMStatistics(std::string isVMStatManager,
-                                                    std::string isTotalAllocationRequired) {
-    std::string vmLevelStatistics;
+long StatisticCollector::getUsedSwapSpace() {
+    long result = getSwapSpace(4);
+    // std::cout << "Used swap space: " + std::to_string(result) << std::endl;
+    return result;
+}
 
-    if (isVMStatManager == "true") {
-        long totalMemoryUsed = getTotalMemoryUsage();
-        double totalCPUUsage = getTotalCpuUsage();
+long StatisticCollector::getTotalSwapSpace() {
+    long result = getSwapSpace(3);
```

Comment:
Why do we pass 3 here?

Reply:
Total swap space value is in the 3rd column

Reply:
Ok